sky/jobs/log_gc.py

```diff
@@ -0,0 +1,187 @@
+        retention_seconds, batch_size=batch_size)
+    for job in jobs:
+        log_file = managed_job_utils.controller_log_file_for_job(job['job_id'])
+        if os.path.exists(log_file):
+            cleaned_at = time.time()
+            ts_str = datetime.fromtimestamp(cleaned_at).strftime(
+                '%Y-%m-%d %H:%M:%S')
+            msg = f'Controller log has been cleaned at {ts_str}.'
+            # Sync down logs will reference to this file directly, so we
+            # keep the file and delete the content.
+            # TODO(aylei): refactor sync down logs if the inode usage
+            # becomes an issue.
+            async with await anyio.open_file(log_file, 'w',
+                                             encoding='utf-8') as f:
+                await f.write(msg + '\n')
+        await managed_job_state.set_controller_logs_cleaned_async(
+            job_id=job['job_id'], logs_cleaned_at=cleaned_at)
+    complete = len(jobs) < batch_size
+    logger.info(f'Cleaned {len(jobs)} controller logs with retention '
+                f'{retention_seconds} seconds, complete: {complete}')
+    return complete
+
+
+async def _clean_task_logs_with_retention(retention_seconds: int,
+                                          batch_size: int = 1000):
+    """Clean task logs with retention.
+
+    Returns:
+        Whether the GC of this round has finished, False means there might
+        still be more task logs to clean.
+    """
+    assert batch_size > 0, 'Batch size must be positive'
+    tasks = await managed_job_state.get_task_logs_to_clean_async(
+        retention_seconds, batch_size=batch_size)
+    for task in tasks:
+        local_log_file = anyio.Path(task['local_log_file'])
+        # We assume the log directory has the following layout:
+        # task-id/
+        #   - run.log
+        #   - tasks/
+        #     - run.log
+        # and also remove the tasks directory on cleanup.
+        task_log_dir = local_log_file.parent.joinpath('tasks')
+        await local_log_file.unlink(missing_ok=True)
+        await context_utils.to_thread(shutil.rmtree,
+                                      str(task_log_dir),
+                                      ignore_errors=True)
+        # We have at least once semantic guarantee for the cleanup here.
+        await managed_job_state.set_task_logs_cleaned_async(
+            spot_job_id=task['spot_job_id'],
+            task_id=task['task_id'],
+            logs_cleaned_at=time.time())
+    complete = len(tasks) < batch_size
+    logger.info(f'Cleaned {len(tasks)} task logs with retention '
+                f'{retention_seconds} seconds, complete: {complete}')
+    return complete
+
+
+@context.contextual_async
+async def run_log_gc():
+    """Run the log garbage collector."""
+    log_dir = os.path.expanduser(managed_job_constants.JOBS_CONTROLLER_LOGS_DIR)
+    os.makedirs(log_dir, exist_ok=True)
+    log_path = os.path.join(log_dir, 'garbage_collector.log')
```

Comment:
Does this log ever get garbage collected? ;)
We can look into it later but it does seem like it could eventually be an issue if you are using 30s retention.